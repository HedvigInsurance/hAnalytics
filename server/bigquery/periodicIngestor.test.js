const {
  start,
  stop,
  addToQueue,
  waitUntilIdle,
  consumeQueue,
} = require("./periodicIngestor");
const inMemoryBackend = require("./periodicIngestorInMemoryBackend");
const setupTable = require("./schema/setupTable");
const createBigQueryConfigMock = require("./config.mock");

test("ingests correctly", async () => {
  const bigQueryConfig = createBigQueryConfigMock();
  start(bigQueryConfig, inMemoryBackend, 100, false, false);

  setupTable(
    "mock_table",
    [
      {
        name: "property",
        type: "STRING",
      },
    ],
    bigQueryConfig
  );

  addToQueue({
    table: "mock_table",
    row: {
      property: "HELLO",
    },
  });

  await waitUntilIdle();
  await stop();

  expect(await consumeQueue()).toMatchSnapshot();
  expect(bigQueryConfig.bigquery.getTables()).toMatchSnapshot();
});

test("ingests exact amount of rows", async () => {
  const bigQueryConfig = createBigQueryConfigMock();
  start(bigQueryConfig, inMemoryBackend, 25, false, false);

  setupTable(
    "mock_table",
    [
      {
        name: "property",
        type: "STRING",
      },
    ],
    bigQueryConfig
  );

  const numberOfRows = 50;

  [...new Array(numberOfRows)].forEach(() => {
    addToQueue({
      table: "mock_table",
      row: {
        property: "HELLO",
      },
    });
  });

  await waitUntilIdle();
  await stop();

  expect(bigQueryConfig.bigquery.getTables()[0].rows.length).toEqual(
    numberOfRows
  );
});

test("doesnt ingest invalid rows", async () => {
  const bigQueryConfig = createBigQueryConfigMock();
  start(bigQueryConfig, inMemoryBackend, 5, false, false);

  setupTable(
    "mock_table",
    [
      {
        name: "property",
        type: "STRING",
      },
    ],
    bigQueryConfig
  );

  const numberOfRows = 25;

  [...new Array(numberOfRows)].forEach(() => {
    addToQueue({
      table: "mock_table",
      row: {
        property: "HELLO",
        context_something_invalid: Math.random(),
      },
    });

    addToQueue({
      table: "mock_table",
      row: {
        property: "HELLO",
      },
    });
  });

  await waitUntilIdle();
  await stop();

  expect(bigQueryConfig.bigquery.getTables()[0].rows.length).toEqual(
    numberOfRows
  );
}, 20000);

test("does ingest if tables update", async () => {
  const bigQueryConfig = createBigQueryConfigMock();
  start(bigQueryConfig, inMemoryBackend, 10, false, false);

  await setupTable(
    "mock_table",
    [
      {
        name: "property",
        type: "STRING",
      },
    ],
    bigQueryConfig
  );

  const numberOfRows = 25;

  [...new Array(numberOfRows)].forEach(() => {
    addToQueue({
      table: "mock_table",
      row: {
        property: "HELLO",
      },
    });

    addToQueue({
      table: "mock_table2",
      row: {
        property_2: "HELLO",
      },
    });
  });

  await setupTable(
    "mock_table2",
    [
      {
        name: "property_2",
        type: "STRING",
      },
    ],
    bigQueryConfig
  );

  await waitUntilIdle();
  await stop();

  expect(bigQueryConfig.bigquery.getTables()[0].rows.length).toEqual(
    numberOfRows
  );
  expect(bigQueryConfig.bigquery.getTables()[1].rows.length).toEqual(
    numberOfRows
  );
}, 20000);

test("does ingest if schema updates", async () => {
  const bigQueryConfig = createBigQueryConfigMock();
  start(bigQueryConfig, inMemoryBackend, 10, false, false);

  await setupTable(
    "embark_track",
    [
      {
        name: "event_id",
        type: "STRING",
      },
    ],
    bigQueryConfig
  );

  const numberOfRows = 30;

  [...new Array(numberOfRows)].forEach((_, index) => {
    addToQueue({
      table: "embark_track",
      row: {
        event_id: "mock_id",
      },
    });

    addToQueue({
      table: "embark_track",
      row: {
        event_id: "mock_id",
        [`property_store_${index * 1000}`]: "HELLO",
        [`property_store_${index * 1000}_double`]: 150,
        [`property_store_${index * 1000}_bool`]: true,
        [`property_store_${index * 1000}_array`]: ["string", "string2"],
      },
    });
  });

  await waitUntilIdle();
  await stop();

  expect(bigQueryConfig.bigquery.getTables()[0].rows.length).toEqual(
    numberOfRows * 2
  );
  expect(bigQueryConfig.bigquery.getTables()).toMatchSnapshot();
}, 20000);
