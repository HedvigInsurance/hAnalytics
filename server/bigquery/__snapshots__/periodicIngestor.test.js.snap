// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does ingest if schema updates 1`] = `
Array [
  Object {
    "config": Object {
      "schema": Object {
        "fields": Array [
          Object {
            "name": "event_id",
            "type": "STRING",
          },
        ],
      },
      "timePartitioning": Object {
        "expirationMS": "7776000000",
        "field": "timestamp",
        "type": "DAY",
      },
    },
    "name": "embark_track",
    "rows": Array [
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_0": "HELLO",
        "property_store_0_array": Array [
          "string",
          "string2",
        ],
        "property_store_0_bool": true,
        "property_store_0_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_1000": "HELLO",
        "property_store_1000_array": Array [
          "string",
          "string2",
        ],
        "property_store_1000_bool": true,
        "property_store_1000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_2000": "HELLO",
        "property_store_2000_array": Array [
          "string",
          "string2",
        ],
        "property_store_2000_bool": true,
        "property_store_2000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_3000": "HELLO",
        "property_store_3000_array": Array [
          "string",
          "string2",
        ],
        "property_store_3000_bool": true,
        "property_store_3000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_4000": "HELLO",
        "property_store_4000_array": Array [
          "string",
          "string2",
        ],
        "property_store_4000_bool": true,
        "property_store_4000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_5000": "HELLO",
        "property_store_5000_array": Array [
          "string",
          "string2",
        ],
        "property_store_5000_bool": true,
        "property_store_5000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_6000": "HELLO",
        "property_store_6000_array": Array [
          "string",
          "string2",
        ],
        "property_store_6000_bool": true,
        "property_store_6000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_7000": "HELLO",
        "property_store_7000_array": Array [
          "string",
          "string2",
        ],
        "property_store_7000_bool": true,
        "property_store_7000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_8000": "HELLO",
        "property_store_8000_array": Array [
          "string",
          "string2",
        ],
        "property_store_8000_bool": true,
        "property_store_8000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_9000": "HELLO",
        "property_store_9000_array": Array [
          "string",
          "string2",
        ],
        "property_store_9000_bool": true,
        "property_store_9000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_10000": "HELLO",
        "property_store_10000_array": Array [
          "string",
          "string2",
        ],
        "property_store_10000_bool": true,
        "property_store_10000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_11000": "HELLO",
        "property_store_11000_array": Array [
          "string",
          "string2",
        ],
        "property_store_11000_bool": true,
        "property_store_11000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_12000": "HELLO",
        "property_store_12000_array": Array [
          "string",
          "string2",
        ],
        "property_store_12000_bool": true,
        "property_store_12000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_13000": "HELLO",
        "property_store_13000_array": Array [
          "string",
          "string2",
        ],
        "property_store_13000_bool": true,
        "property_store_13000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_14000": "HELLO",
        "property_store_14000_array": Array [
          "string",
          "string2",
        ],
        "property_store_14000_bool": true,
        "property_store_14000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_15000": "HELLO",
        "property_store_15000_array": Array [
          "string",
          "string2",
        ],
        "property_store_15000_bool": true,
        "property_store_15000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_16000": "HELLO",
        "property_store_16000_array": Array [
          "string",
          "string2",
        ],
        "property_store_16000_bool": true,
        "property_store_16000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_17000": "HELLO",
        "property_store_17000_array": Array [
          "string",
          "string2",
        ],
        "property_store_17000_bool": true,
        "property_store_17000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_18000": "HELLO",
        "property_store_18000_array": Array [
          "string",
          "string2",
        ],
        "property_store_18000_bool": true,
        "property_store_18000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_19000": "HELLO",
        "property_store_19000_array": Array [
          "string",
          "string2",
        ],
        "property_store_19000_bool": true,
        "property_store_19000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_20000": "HELLO",
        "property_store_20000_array": Array [
          "string",
          "string2",
        ],
        "property_store_20000_bool": true,
        "property_store_20000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_21000": "HELLO",
        "property_store_21000_array": Array [
          "string",
          "string2",
        ],
        "property_store_21000_bool": true,
        "property_store_21000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_22000": "HELLO",
        "property_store_22000_array": Array [
          "string",
          "string2",
        ],
        "property_store_22000_bool": true,
        "property_store_22000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_23000": "HELLO",
        "property_store_23000_array": Array [
          "string",
          "string2",
        ],
        "property_store_23000_bool": true,
        "property_store_23000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_24000": "HELLO",
        "property_store_24000_array": Array [
          "string",
          "string2",
        ],
        "property_store_24000_bool": true,
        "property_store_24000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_25000": "HELLO",
        "property_store_25000_array": Array [
          "string",
          "string2",
        ],
        "property_store_25000_bool": true,
        "property_store_25000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_26000": "HELLO",
        "property_store_26000_array": Array [
          "string",
          "string2",
        ],
        "property_store_26000_bool": true,
        "property_store_26000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_27000": "HELLO",
        "property_store_27000_array": Array [
          "string",
          "string2",
        ],
        "property_store_27000_bool": true,
        "property_store_27000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_28000": "HELLO",
        "property_store_28000_array": Array [
          "string",
          "string2",
        ],
        "property_store_28000_bool": true,
        "property_store_28000_double": 150,
      },
      Object {
        "event_id": "mock_id",
      },
      Object {
        "event_id": "mock_id",
        "property_store_29000": "HELLO",
        "property_store_29000_array": Array [
          "string",
          "string2",
        ],
        "property_store_29000_bool": true,
        "property_store_29000_double": 150,
      },
    ],
  },
]
`;

exports[`does keep invalid rows 1`] = `Array []`;

exports[`ingests correctly 1`] = `Array []`;

exports[`ingests correctly 2`] = `
Array [
  Object {
    "config": Object {
      "schema": Object {
        "fields": Array [
          Object {
            "name": "property",
            "type": "STRING",
          },
        ],
      },
      "timePartitioning": Object {
        "expirationMS": "7776000000",
        "field": "timestamp",
        "type": "DAY",
      },
    },
    "name": "mock_table",
    "rows": Array [
      Object {
        "property": "HELLO",
      },
    ],
  },
]
`;
